File: day12-bs-tree/bstree.c
// A Binary Search Tree in C.

#include <stdio.h>
#include <stdlib.h>
#include "bstree.h"
#include "intlist.h"

Tree * new_tree() {
  // Creates a new, empty Tree.
  Tree *tree = NULL;

  tree = malloc(sizeof(Tree));

  if(tree != NULL) 
    tree->root = NULL;

  return tree;
}

Tree * new_tree_from_list(IntList *list) {
  // Given an IntList, creates a Tree from it.
  Tree *tree = NULL;
  int count;

  if(list != NULL) {
    tree = new_tree();

    if(tree != NULL) {
      // Insert every item in 'list' into the Tree.
      for(count = 0; count < list->size; count++)
	insert(tree, list->list[count]);
    }
  }

  return tree;
}

int contains(Tree *tree, int item) {
  // Determines if 'item' is in the given Tree.
  int result = 0;

  if(tree != NULL)
    result = _contains(tree->root, item);
  
  return result;    
}

int _contains(Leaf *leaf, int item) {
  // Uses recursion to test for containment.
  int result;

  if(leaf == NULL)
    return 0;

  if(item == leaf->data)
    result = 1;
  else if(item < leaf->data)
    result = _contains(leaf->left, item);
  else
    result = _contains(leaf->right, item);

  return result;
}

void insert(Tree *tree, int item) {
  // Inserts an item into the given Tree.
  if(tree != NULL) {
    tree->root = _insert(tree->root, item);
  }
}

Leaf * _insert(Leaf *leaf, int item) {
  // Inserts the new item via recursion.
  if(leaf == NULL)
    return new_leaf(item);

  if(item < leaf->data)
    leaf->left = _insert(leaf->left, item);
  else
    leaf->right = _insert(leaf->right, item);

  return leaf;
}

void print_tree(Tree *tree) {
  // Figured out a great way to print trees.
  if(tree != NULL)
    _print_tree(tree->root, 0);
}

void _print_tree(Leaf *leaf, int indent) {
  // Prints the tree opening to the right via recursion.
  if(leaf != NULL) {
    _print_tree(leaf->right, indent + 2);
    print_indent(indent);
    printf("%d\n", leaf->data);
    _print_tree(leaf->left, indent + 2);
  }
}

void print_indent(int indent) {
  // Prints a bunch of whitespace.
  int count;

  for(count = 0; count < indent; count++)
    printf(" ");
}

Leaf * new_leaf(int item) {
  Leaf *leaf = NULL;

  leaf = malloc(sizeof(Leaf));

  if(leaf != NULL) {
    leaf->data = item;
    leaf->left = NULL;
    leaf->right = NULL;
  }
  
  return leaf;
}

File: day12-bs-tree/bs_tree.py
# A binary search tree in Python.

class BSTree():
    '''A binary search tree.'''
    def __init__(self, items=None):
        self.root = None
        if items:  # Can be initialized with a list.
            for item in items:
                self.insert(item)

    def __str__(self):
        '''A pretty picture of a tree.
        May this be the definitive way to print trees, ever. EVER.
        Don't need a special case for empty trees, either :)
        '''
        leaves = []
        self._str(leaves, self.root, 0)
        return ''.join(leaves).rstrip()

    def _str(self, leaves, leaf, indent):
        '''Never have I high-fived myself so hard.
        This generates a tree string expanding out to the right via recursion.
        The indent of each leaf on its line its carried through the recursive
        process, increasing for each layer.'''
        if leaf:
            self._str(leaves, leaf.right, indent + 2)  # Arbitrary pad of 2.
            leaves.append(''.join((' ' * indent, str(leaf), '\n')))
            self._str(leaves, leaf.left, indent + 2)

    def __contains__(self, item):
        '''Determines if a value is in the tree.'''
        return self._contains(self.root, item)

    def _contains(self, leaf, item):
        '''Uses recursion to test for containment.'''
        if not leaf:
            return False
        if item == leaf.data:
            result = True
        elif item < leaf.data:
            result = self._contains(leaf.left, item)
        else:
            result = self._contains(leaf.right, item)
        return result

    def insert(self, item):
        '''Insert a value into the tree.'''
        self.root = self._insert(self.root, item)

    def _insert(self, leaf, item):
        '''Inserts a new Leaf via recursion.'''
        if not leaf:
            return Leaf(item)
        if item < leaf.data:
            leaf.left = self._insert(leaf.left, item)
        else:
            leaf.right = self._insert(leaf.right, item)
        return leaf

class Leaf():
    '''Or 'Node', if you wish.'''
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return str(self.data)
        
